### PR TITLE
Refactor User services

### PR DESCRIPTION
**Pull request by Nitemaeric**
## Context

To prepare for [#301 Remove School Users](https://github.com/DFE-Digital/itt-mentor-services/pull/301), we should align class naming conventions of [Service Objects](https://github.com/DFE-Digital/itt-mentor-services/blob/main/docs/conventions/service_objects.md).

## Changes proposed in this pull request

- Refactor `UserInviteService.call(user, organisation)` to `User::Invite.call(user:, organisation:)`
- Refactor `RemoveUserService.call(user, organisation)` to `User::Remove.call(user:, organisation:)`

## Guidance to review

- Ensure refactor aligns with our Service Object conventions.
- Specs all pass.
- Review commit messages.

### CODE REVIEW

File: config/locales/en/user_mailer.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+en:
+  user_mailer:
+    user_invitation_notification:
+      subject: You have been invited to %{organisation_name}
+      body: |
```
[the remark this answers is not in the document]
Comment (Nitemaeric) [reply config/locales/en/user_mailer.yml:5]: I agree. I had the same initial reaction when we started adding emails.

We don't have any complex email templates as of now, just a few lines of text, but when we require any styling, we'll definitely need to extract the body into template files.

File: app/mailers/user_mailer.rb [hunk 2]
```diff
@@ -1,30 +1,13 @@
...
-    mailer_options = {
-      to: user.email,
-      subject: t("#{translation_root}.subject", organisation_name: organisation.name),
-      body: t("#{translation_root}.body",
-              user_name: user.full_name,
-              organisation_name: organisation.name),
-    }
-
-    notify_email(mailer_options)
+  def user_invitation_notification(user, organisation)
```
Comment (JamieCleare2525) [inline app/mailers/user_mailer.rb:2]: Surely `user` is self explanatory here.

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -1,30 +1,13 @@
...
-             sign_in_url:)
-
-    mailer_options = {
-      to: user.email,
-      subject: t("#{translation_root}.subject", organisation_name: organisation.name),
-      body:,
-    }
-
-    notify_email(mailer_options)
+  def user_removal_notification(user, organisation)
```
Comment (JamieCleare2525) [inline app/mailers/user_mailer.rb:8]: Same as above.

File: config/locales/en/user_mailer.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nitemaeric) [reply config/locales/en/user_mailer.yml:5]: :/ That's a shame.

File: app/mailers/user_mailer.rb (hunk 2 above)
Comment (Nitemaeric) [reply app/mailers/user_mailer.rb:2]: This change is based on [this message](https://ukgovernmentdfe.slack.com/archives/C06962474AC/p1705488689303999) - [Direct Link to article](https://katanacode.com/blog/ruby-on-rails-action-mailer-how-to-name-mailer-methods/).

It _could_ be made implicit that the resource being affected is the user themself, but to keep things scalably explicit, prefixing with `user_` identifies that a user is the resource being effected.

For example, notifications about successfully submitted claims would be `UserMailer#claim_submitted_notification`.

Comment (Nitemaeric) [reply app/mailers/user_mailer.rb:2]: Hmm, that's true. I'll change and add a conventions doc.

Comment (Nitemaeric) [reply app/mailers/user_mailer.rb]: Resolved in https://github.com/DFE-Digital/itt-mentor-services/pull/303/commits/5aa71ade825ef6452df26df71a5522b86cdbd56d and https://github.com/DFE-Digital/itt-mentor-services/pull/303/commits/6080bdff0f1e3735847653cdc896f38a8f6b7178 😄

**Review by JamieCleare2525 — APPROVED** [review]

File: docs/conventions/mailers.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+>
+> To develop a naming convention, we must first reflect on how our code is being used, and look for some common patterns that apply to most use cases. In doing so, I’ve realised that most mailers can all be boiled down to these common features:
+>
+> #### Recipient
+>
+> The recipient is the entity that will receive the email. This will normally be a person, but it could also be another app or remote service that receives email inputs. Recipients might be your service users, customers, internal admins, or unauthenticated visitors to your website.
+>
+> ### Resource
+>
+> Emails are usually sent out to provide information about a specific resource or domain concept within the system. In other words, emails are usually sent about something relevant to some thing. An article has been published, a subscription has lapsed, a purchase has been completed. These will often be the models that you’ve defined in your code, but they might also be more abstract concepts (like a password reminder).
```
Comment (CatalinVoineag) [inline docs/conventions/mailers.md:30]: Should we include something here about the need to add an `unsubscribe` button to these types of emails?

**Review by CatalinVoineag — APPROVED** [review]
Just one comment, not really a blocker. This looks very good

File: docs/conventions/mailers.md (hunk 4 above)
Comment (Nitemaeric) [reply docs/conventions/mailers.md:30]: We only need unsubscribe links for marketing emails. Transactional emails do not need unsubscribe links. 😄

**Pull request merged**